dids/did_dht.go

```diff
@@ -0,0 +1,310 @@
+package dids
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"golang.org/x/net/dns/dnsmessage"
+
+	"github.com/tbd54566975/web5-go/crypto/dsa"
+	"github.com/tv42/zbase32"
+)
+
+var txtEntityNames = map[string]struct{}{
+	"vm":   {},
+	"auth": {},
+	"asm":  {},
+	"agm":  {},
+	"inv":  {},
+	"del":  {},
+}
+
+var relationshipDNStoDID = map[string]string{
+	"auth": "authentication",
+	"asm":  "assertionMethod",
+	"agm":  "keyAgreement",
+	"inv":  "capabilityInvocation",
+	"del":  "capabilityDelegation",
+}
+
+var keyTypes = map[string]string{
+	"0": dsa.AlgorithmIDED25519,
+	"1": dsa.AlgorithmIDSECP256K1,
+	"2": dsa.AlgorithmIDSECP256K1,
+}
+
+// dhtDIDRecord is used to structure the DNS representation of a DID
+type dhtDIDRecord struct {
+	rootRecord string
+	records    map[string]string
+}
+
+func (rec *dhtDIDRecord) DIDDocument(didURI string) *Document {
+	relationshipMap := parseVerificationRelationships(rec.rootRecord)
+
+	// Now we have a did in a dns record. yay
+	document := &Document{
+		ID: didURI,
+	}
+
+	// Now create the did document
+	for name, data := range rec.records {
+
+		switch {
+		case strings.HasPrefix(name, "_k"):
+			{
+				vMethod, err := UnmarshalVerificationMethod(data)
+				if err != nil {
+					// TODO handle error
+				}
+
+				// extracting kN from _kN._did
+				entryId := strings.Split(name, ".")[0][1:]
+				relationships, ok := relationshipMap[entryId]
+
+				if !ok {
+					// no relationships
+					continue
+				}
+
+				opts := []string{}
+				for _, r := range relationships {
+					if o, ok := relationshipDNStoDID[r]; ok {
+						opts = append(opts, o)
+					}
+				}
+
+				document.AddVerificationMethod(
+					*vMethod,
+					Purposes(opts...),
+				)
+			}
+		case strings.HasPrefix(name, "_s"):
+			s := UnmarshalService(data)
+			document.AddService(s)
+		case strings.HasPrefix(name, "_cnt"):
+			// TODO add controller https://did-dht.com/#controller
+			// optional field
+			// comma-separated list of controller DID identifiers
+			document.Controller = strings.Split(data, ",")
+		case strings.HasPrefix(name, "_aka"):
+			// TODO add aka https://did-dht.com/#also-known-as
+			document.AlsoKnownAs = strings.Split(data, ",")
+		default:
+		}
+	}
+
+	return document
+}
+
+func ResolveDIDDHT(uri, relay string, client *http.Client) (*Document, error) {
```

Comment:
i think we'll want to return a `ResolutionResult` instead of `Document` like [so](https://github.com/TBD54566975/web5-go/blob/main/dids/didjwk.go#L78) to remain consistent with the [did core spec](https://www.w3.org/TR/did-core/#did-resolution) and our other SDKs!

Reply:
indeed. I'll add this update